ir/src/passes/constant_propagation.rs

```diff
@@ -0,0 +1,36 @@
+use std::ops::ControlFlow;
+
+use air_pass::Pass;
+use miden_diagnostics::DiagnosticsHandler;
+
+use crate::MirGraph;
+
+pub struct ConstantPropagation<'a> {
```

Comment:
This pass becomes trivial with the design I outlined above, after one has implemented the generic pre- and post-order visitors for `Operation`, basically:

1. Perform a post-order visit of the `Function`, pushing all of the operations which have at least one use of a constant op result on a stack. This could also be an opportunity to unique constants, and separately enqueue constant ops that should be erased, while also recording a mapping from the values produced by those ops to the canonical value for that constant.
2. Process the next item on the work stack, and determine if that operation can be constant folded based on its constant operands. If so, fold the operation. If folded in place, add all users of the op's results to the work stack, and go to the next op on the stack. If the op can be folded away completely and replaced with a constant value, go to step 3. If the operation could not be folded, start processing the next item.
3. Materialize a new constant op representing the folded constant, and insert it before the op we want to replace.
4. Replace the results of the original op with those of the constant op (updating all the users at the same time). Add each unique using operation to the work stack, if not already present on the stack.
5. Erase the original op. Go to step 2.

Because the data flow graph is implicit in the edges between operations/operands/results, doing the above is quite efficient, since we only need to do a single traversal of the graph, and we only revisit operations when one of their uses has been constant folded (so as to see if the change makes it foldable).

Reply:
We agree.